### PR TITLE
CUDA - LinearAssemble*Diagonal

### LINKED ISSUE #425: AssembleLinearQF/Op Fallback
The `AssembleLinearQF/Op` fallback is failing. I want to provide a basic version of `CeedAssembleLinearQFunction`, `CeedAssembleLinearOperator`, and `CeedOperatorCreateFDMElementInverse` to get some preconditioning functionality for all backends.

Unfortunately, the CUDA backends seem to permute their vectors in some fashion, perhaps by blocksize? It looks like we have some sort of permutation along the lines of [elem1field1, elem2field1, elem1field2, elem2field2, ...] rather than [elem1field1, elem1field2, elem2field1, elem2field2, ...]. Am I reading that right? Is there a way to get the 'canonical' ordering from all of our diagrams/tests?

 CUDA grad qdata vector: (16 qpts per element)
```
CeedVector length 288
    0.18150449
    0.34027778
    0.34027778
    0.18150449
    0.34027778
    0.63793995
    0.63793995
    0.34027778
    0.34027778
    0.63793995
    0.63793995
    0.34027778
    0.18150449
    0.34027778
    0.34027778
    0.18150449
    0.18150449
    0.34027778
    0.34027778
    0.18150449
    0.34027778
    0.63793995
    0.63793995
    0.34027778
    0.34027778
    0.63793995
    0.63793995
    0.34027778
    0.18150449
    0.34027778
    0.34027778
    0.18150449
...
```
CPU serial qdata vector:
```
CeedVector length 288
    0.18150449
    0.34027778
    0.34027778
    0.18150449
    0.34027778
    0.63793995
    0.63793995
    0.34027778
    0.34027778
    0.63793995
    0.63793995
    0.34027778
    0.18150449
    0.34027778
    0.34027778
    0.18150449
    0.08066866
    0.15123457
    0.15123457
    0.08066866
    0.15123457
    0.28352887
    0.28352887
    0.15123457
    0.15123457
    0.28352887
    0.28352887
    0.15123457
    0.08066866
    0.15123457
    0.15123457
    0.08066866
...
```

### PR DESCRIPTION
**Pull request by jeremylt**
~~Initial work for~~ diagonal assembly for CUDA/MAGMA backends.

Closes #425

### CODE REVIEW

**Comment by jeremylt** [comment]
Work in progress. Currently I get failures for all backends execpt `/gpu/cuda/gen`, which is strange
```
not ok 31 t533-operator /gpu/cuda/ref
ok 32 t533-operator /gpu/cuda/ref stdout
not ok 33 t533-operator /gpu/cuda/ref stderr
# +/home/jeremy/Dev/libCEED/backends/cuda/ceed-cuda-vector.c:41 in CeedSyncH2D_Cuda(): CUDA_ERROR_INVALID_VALUE
not ok 34 t533-operator /gpu/cuda/reg
ok 35 t533-operator /gpu/cuda/reg stdout
not ok 36 t533-operator /gpu/cuda/reg stderr
# +/home/jeremy/Dev/libCEED/backends/cuda/ceed-cuda-vector.c:41 in CeedSyncH2D_Cuda(): CUDA_ERROR_INVALID_VALUE
not ok 37 t533-operator /gpu/cuda/shared
ok 38 t533-operator /gpu/cuda/shared stdout
not ok 39 t533-operator /gpu/cuda/shared stderr
# +/home/jeremy/Dev/libCEED/backends/cuda/ceed-cuda-vector.c:41 in CeedSyncH2D_Cuda(): CUDA_ERROR_INVALID_VALUE
ok 40 t533-operator /gpu/cuda/gen
ok 41 t533-operator /gpu/cuda/gen stdout
ok 42 t533-operator /gpu/cuda/gen stderr
not ok 43 t533-operator /gpu/magma
ok 44 t533-operator /gpu/magma stdout
not ok 45 t533-operator /gpu/magma stderr
# +/home/jeremy/Dev/libCEED/backends/cuda/ceed-cuda-vector.c:41 in CeedSyncH2D_Cuda(): CUDA_ERROR_INVALID_VALUE
```

**Comment by jeremylt** [comment]
Bug fixed. We now have an initial capability to build the diagonal on the host.

**Comment by jeremylt** [comment]
Diagonal assembly on the device is all set.

Current Full test suite failures:
```
$ OCCA_DIR= make prove-all -j8 PROVE_OPTS=-v
Testing backends: /cpu/self/ref/serial /cpu/self/ref/blocked /cpu/self/opt/serial /cpu/self/opt/blocked /cpu/self/memcheck/serial /cpu/self/memcheck/blocked /cpu/self/avx/serial /cpu/self/avx/blocked /cpu/self/xsmm/serial /cpu/self/xsmm/blocked /gpu/cuda/ref /gpu/cuda/reg /gpu/cuda/shared /gpu/cuda/gen /gpu/magma /cpu/self/tmpl /cpu/self/tmpl/sub
...
Test Summary Report
-------------------
petsc-area            (Wstat: 0 Tests: 45 Failed: 2)
  Failed tests:  37, 39
petsc-bpsraw          (Wstat: 0 Tests: 45 Failed: 6)
  Failed tests:  31, 33, 37, 39, 43, 45
petsc-bpssphere       (Wstat: 0 Tests: 45 Failed: 2)
  Failed tests:  37, 39
petsc-multigrid       (Wstat: 0 Tests: 45 Failed: 10)
  Failed tests:  31, 33-34, 36-37, 39-40, 42-43, 45
fluids-navierstokes   (Wstat: 0 Tests: 135 Failed: 3)
  Failed tests:  41, 86, 131
```
`petsc-area` and `petsc-bpssphere`, known issue with `/gpu/cuda/shared` #462 
`petsc-bpsraw`, known issue with `/gpu/cuda/ref', '/gpu/cuda/shared', and '/gpu/magma' #462 
`fluids-navierstokes`, known issue discussed on Slack
`petsc-multigrid`, debugging now

**Comment by jeremylt** [comment]
Update: failing was due to missing `&`. PR is ready for review. Now `petsc-mulitgrid` and `solids-elasticity` run with the GPU backends (though I need to do a separate PR to make the solids example use GPU vectors)

**Comment by valeriabarra** [comment]
For the `fluids-navierstokes` one I had left a [comment](https://github.com/CEED/libCEED/pull/540#issuecomment-646176963). Would [this commit](https://github.com/CEED/libCEED/pull/540/commits/3e1acbe8071035eb7a41fcb16bd5cbe3ee128268) fix it?

**Comment by jeremylt** [comment]
@valeriabarra with that commit cherry-picked I get
```
# +[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
# +[0]PETSC ERROR: Floating point exception
# +[0]PETSC ERROR: Infinite or not-a-number generated in norma
# +[0]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
# +[0]PETSC ERROR: Petsc Development GIT revision: v3.13.1-182-gd9ef940ef5  GIT Date: 2020-05-15 09:38:31 -0500
# +[0]PETSC ERROR: build/fluids-navierstokes on a petsc-opt named sinensis by jeremy Tue Jun 23 14:59:31 2020
# +[0]PETSC ERROR: Configure options --download-fblaslapack --download-parmetis --download-metis --download-zlib --download-hdf5 --download-netcdf --download-pnetcdf --download-exodusii --with-cuda --download-viennacl --COPTFLAGS="-O3 -march=native" -CXXOPTFLAGS="-O3 -march=native" FOPTFLAGS="-O3 -march=native" --CUDAOPTFLAGS=-O3 --with-debugging=0 -j8
# +[0]PETSC ERROR: #1 TSErrorWeightedNorm2() line 5853 in /home/jeremy/Dev/petsc/src/ts/interface/ts.c
# +[0]PETSC ERROR: #2 TSErrorWeightedNorm() line 6039 in /home/jeremy/Dev/petsc/src/ts/interface/ts.c
# +[0]PETSC ERROR: #3 TSAdaptChoose_Basic() line 31 in /home/jeremy/Dev/petsc/src/ts/adapt/impls/basic/adaptbasic.c
# +[0]PETSC ERROR: #4 TSAdaptChoose() line 920 in /home/jeremy/Dev/petsc/src/ts/adapt/interface/tsadapt.c
# +[0]PETSC ERROR: #5 TSStep_RK() line 810 in /home/jeremy/Dev/petsc/src/ts/impls/explicit/rk/rk.c
# +[0]PETSC ERROR: #6 TSStep() line 3721 in /home/jeremy/Dev/petsc/src/ts/interface/ts.c
# +[0]PETSC ERROR: #7 TSSolve() line 4127 in /home/jeremy/Dev/petsc/src/ts/interface/ts.c
# +[0]PETSC ERROR: #8 main() line 1366 in /home/jeremy/Dev/libCEED/examples/fluids/navierstokes.c
# +[0]PETSC ERROR: PETSc Option Table entries:
# +[0]PETSC ERROR: -ceed /gpu/cuda/gen
# +[0]PETSC ERROR: -center 62.5,62.5,187.5
# +[0]PETSC ERROR: -degree 3
# +[0]PETSC ERROR: -dm_plex_box_faces 1,1,2
# +[0]PETSC ERROR: -ksp_atol 1e-4
# +[0]PETSC ERROR: -ksp_rtol 1e-3
# +[0]PETSC ERROR: -ksp_type bcgs
# +[0]PETSC ERROR: -lx 125
# +[0]PETSC ERROR: -ly 125
# +[0]PETSC ERROR: -lz 250
# +[0]PETSC ERROR: -rc 100.
# +[0]PETSC ERROR: -snes_atol 1e-3
# +[0]PETSC ERROR: -snes_lag_jacobian 100
# +[0]PETSC ERROR: -snes_lag_jacobian_persists
# +[0]PETSC ERROR: -snes_mf_operator
# +[0]PETSC ERROR: -test explicit
# +[0]PETSC ERROR: -thetaC -35.
# +[0]PETSC ERROR: -ts_dt 1e-3
# +[0]PETSC ERROR: -units_kilogram 1e-9
# +[0]PETSC ERROR: ----------------End of Error Message -------send entire error message to petsc-maint@mcs.anl.gov----------
# +--------------------------------------------------------------------------
# +MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_SELF
# +with errorcode 1366072.
# +
# +NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
# +You may or may not see output from other processes, depending on
# +exactly when Open MPI kills them.
# +--------------------------------------------------------------------------
```

**Comment by valeriabarra** [comment]
Oh nooo 😓 never mind then

File: backends/cuda/ceed-cuda-operator.c [hunk 1]
```diff
@@ -649,25 +668,515 @@ static int CeedOperatorLinearAssembleQFunction_Cuda(CeedOperator op,
...
+  // CEED_EVAL_INTERP
+  ierr = CeedBasisGetInterp(basisin, &interpin); CeedChk(ierr);
+  ierr = cudaMalloc((void **)&diag->interpin, iBytes); CeedChk_Cu(ceed, ierr);
+  ierr = cudaMemcpy(diag->interpin, interpin, iBytes,
+                    cudaMemcpyHostToDevice); CeedChk_Cu(ceed, ierr);
+  ierr = CeedBasisGetInterp(basisout, &interpout); CeedChk(ierr);
+  ierr = cudaMalloc((void **)&diag->interpout, iBytes); CeedChk_Cu(ceed, ierr);
+  ierr = cudaMemcpy(diag->interpout, interpout, iBytes,
+                    cudaMemcpyHostToDevice); CeedChk_Cu(ceed, ierr);
+
```
Comment (nbeams) [inline backends/cuda/ceed-cuda-operator.c:1022]: Following the usual cuda backend notation, I think the `identity`, `interpin/out` and `gradin/out` members of `diag` should be named `d_indentity`, `d_interpin`, etc. to indicate that they are on the device.

Comment (jeremylt) [reply backends/cuda/ceed-cuda-operator.c:1022]: Agreed and updated. Thank you.

**Comment by jeremylt** [comment]
And with the new leak fixes in master this branch runs clean (modulo the typical libcuda stuff, but that's not us)

**Pull request merged**